<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>KnockoffGWAS tutorial</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">KnockoffGWAS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="tutorial.html">
    <span class="fa fa-university"></span>
     
    Tutorial
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-chalkboard-teacher"></span>
     
    Talks
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="talks/cgsi-2022/cgsi-2022.slides.html">CGSI 2022</a>
    </li>
  </ul>
</li>
<li>
  <a href="ukbiobank.html">
    <span class="fa fa-dna"></span>
     
    UK Biobank
  </a>
</li>
<li>
  <a href="https://github.com/msesia/knockoffgwas">
    <span class="fa fa-github"></span>
     
    Software
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="mailto:sesia@marshall.usc.edu">
    <span class="fa fa-envelope-o"></span>
     
    Contact
  </a>
</li>
<li>
  <a href="https://msesia.github.io">
    <span class="fa fa-address-card"></span>
     
    About me
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">KnockoffGWAS tutorial</h1>

</div>


<head>
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-110810441-3"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-110810441-3');
</script>
</head>
<p>A toy genetic dataset containing 1000 artificial samples typed at 2000 loci (divided between chromosome 21 and 22) is available from the <a href="https://github.com/msesia/knockoffgwas">software repository</a> as a tutorial for <em>KnockoffGWAS</em>. This data set contains both genotypes and phased haplotypes for all individuals, stored in the standard BED and BGEN formats accepted as input by <em>KnockoffGWAS</em>, along with the phenotype information and all necessary auxiliary files. Note that this tutorial is intended to be run on a Linux system.</p>
<p style="text-align:center;">
<img src="images/data.png" alt="Directory tree structure of tutorial data set"
    title="Directory tree structure of tutorial data set" width="60%" height="60%" />
<figcaption style="text-align: center">
<small>Directory tree structure of tutorial data set.</small>
</figcaption>
</p>
<p>To run the tutorial, download the software repository and enter the corresponding main directory. Then, compile the <em>snpknock2</em> C++ program (to generate the knockoffs) by typing the following commands in your terminal:</p>
<pre><code>cd snpknock2
make
cd ..</code></pre>
<p>After a successful compilation, you can execute the script <code>analyze.sh</code> contained therein.</p>
<pre><code>./analyze.sh</code></pre>
<p>This will first verify whether the system dependencies (plink, R, and the snpknock2 executable compiled above), and then install any missing required R packages. Then, it will carry out an entire association analysis which should take approximately 5 minutes on a personal computer from start to finish. This analysis consists of 4 main modules, each encoded by a distinct script located within the sub-directory <code>knockoffgwas/</code>, and all automatically coordinated by <code>analyze.sh</code>. These modules organize the analysis as follows:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Partitions.</strong> Partition the available SNPs into contiguous groups at different levels of resolution, based on the genetic distance information provided with the data and the desired resolution values (measured in cM) specified by the user (a default option is specified in <code>knockoffgwas/module_1_partition.sh</code>). The genome partitions computed by the module will be stored in the intermediate result directory <code>tmp/partitions</code>.</p></li>
<li><p><strong>Knockoffs.</strong> Generate knockoff genotypes for all specified genome partitions. This operation is carried out by the script <code>knockoffgwas/module_2_knockoffs.sh</code>, which executes the <em>snpknock2</em> program compiled above. The output knockoffs will be stored in standard BED format within the intermediate result directory <code>tmp/knockoffs</code>, merged with a copy of the original genotypes for subsequent convenience. A separate set of BED and auxiliary files will be produced for each available chromosome. The second part of this module, <code>knockoffgwas/module_2b_knockoffs_god.sh</code>, computes goodness-of-fit-diagnostics for the knockoffs corresponding to all chromosomes and resolutions. Plots summarizing the goodness-of-fit-diagnostics are stored in the directory <code>tmp/knockoffs_gof_plots/</code>. Thse goodness-of-fit-diagnostics should reflect the correct exchangeability of the knockoffs with the original genotypes; see the screenshot below for an example on the tutorial data, and the accompanying paper for more details.</p></li>
<li><p><strong>Statistics.</strong> Compute knockoff test statistics for all groups of SNPs, separately for each specified genome partition. This operation is carried out by the script <code>knockoffgwas/module_3_statistics.sh</code>, which first merges the knockoff-augmented genotype files for all chromosomes and then converts them into a large <a href="https://privefl.github.io/bigsnpr/">file-backed matrix</a>. Finally, this script fits a sparse generalized linear regression model to predict the phenotype of interest given the knockoff-augmented data using a very <a href="https://privefl.github.io/bigstatsr/">efficient implementation of the lasso</a>, and utilizes that to compute the final knockoff test statistics, separately for each specified genome partition.</p></li>
<li><p><strong>Filter.</strong> The last module applies the knockoff filter to the statistics computed above, producing a separate list of discoveries for each specified genome partition. This operation is carried out by the script <code>knockoffgwas/module_4_discover.sh</code>. The default FDR level is 10%, but it can be modified within this script. The final lists of discoveries will be written in the directory <code>results/</code>.</p></li>
</ol>
<p style="text-align:center;">
<img src="images/analyze.png" alt="Screenshot of KnockoffGWAS analysis on tutorial data"
    title="KnockoffGWAS analysis on tutorial data" width="65%" height="65%" />
<figcaption style="text-align: center">
<small>Screenshot of KnockoffGWAS analysis on tutorial data.</small>
</figcaption>
</p>
<p style="text-align:center;">
<img src="images/knockoffs_chr21_res5.png" alt="Screenshot of some goodness-of-fit diagnostics for knockoffs on tutorial data"
    title="Goodness-of-fit diagnostics for knockoffs on tutorial data." width="65%" height="65%" />
<figcaption style="text-align: center">
<small>Goodness-of-fit diagnostics for knockoffs on tutorial data.</small>
</figcaption>
</p>
<p>The final results can be visualized interactively with the script <code>visualize.sh</code>, which will launch a <a href="https://shiny.rstudio.com/">Shiny</a> app in your browser. Some additional R packages are required by the visualization tool, and will be automatically installed if not found.</p>
<pre><code>./visualize.sh</code></pre>
<p style="text-align:center;">
<img src="images/visualize.png" alt="Screenshot of KnockoffGWASView  on tutorial data"
    title="Script to visualize results of KnockoffGWAS analysis on tutorial data" width="65%" height="65%" />
</p>
<p>The expected results for the analysis of this toy dataset are provided in the directory <code>results/</code> and can be visualized through interactive Chicago plots by running the script <code>visualize.sh</code> before running <code>analyze.sh</code>. Note that the script <code>analyze.sh</code> will overwrite the default results.</p>
<p style="text-align:center;">
<img src="images/view_1.png" alt="Screenshot of full-genome visualization of KnockoffGWAS discoveries on tutorial data"
    title="Full-genome visualization of KnockoffGWAS discoveries on tutorial data" width="100%" height="100%" />
<figcaption style="text-align: center">
<small>Full-genome visualization of KnockoffGWAS discoveries on tutorial data.</small>
</figcaption>
</p>
<p style="text-align:center;">
<img src="images/view_2.png" alt="Screenshot of single-chromosome visualization of KnockoffGWAS discoveries on tutorial data"
    title="Single-chromosome visualization of KnockoffGWAS discoveries on tutorial data" width="100%" height="100%" />
<figcaption style="text-align: center">
<small>Single-chromosome visualization of KnockoffGWAS discoveries on tutorial data.</small>
</figcaption>
</p>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
